Task 4

enginire4ek opened the pull request:
- The application is designed according to the new architecture "Clean Architecture" and is divided into 4 modules: app, domain, presentation, data
- Added Dagger 2 framework for compile-time dependency injection.
- Handled an error with no internet connection by redirecting to the error fragment
- Changed error display on profile page

enginire4ek committed:
Task 4
- The application is designed according to the new architecture "Clean Architecture" and is divided into 4 modules: app, domain, presentation, data
- Added Dagger 2 framework for compile-time dependency injection.
- Handled an error with no internet connection by redirecting to the error fragment
- Changed error display on profile page

enginire4ek committed:
Task 4. Updates
- UserSignUp.kt moved from data-layer to domain-layer

102Rudy commented on app/src/main/java/com/example/pinterbest/App.kt:
```diff
@@ -0,0 +1,23 @@
+
+    private lateinit var appComponent: ApplicationComponent
+
+    override fun getApplicationAPI(): ApplicationAPI {
```
I suggest to keep `camelCase` even for abbreviations (`ApplicationApi`)

enginire4ek replied:
Fixed

102Rudy commented on app/src/main/java/com/example/pinterbest/di/ApplicationComponent.kt:
```diff
@@ -0,0 +1,22 @@
+@Singleton
+@Component(
+    modules = [
+        AppModule::class, PinsModule::class,
+        ViewModelModule::class, AuthModule::class, ProfileModule::class, SessionModule::class
```
nit: this may be cleaner if rewrite each module on new line, like

```kotlin
modules = [
  AppModule::class,
  PinsModule::class,
  ViewModelModule::class,
  AuthModule::class,
  ProfileModule::class,
  SessionModule::class,
```

enginire4ek replied:
Fixed

102Rudy commented on app/src/main/java/com/example/pinterbest/di/PinsModule.kt:
```diff
@@ -0,0 +1,18 @@
+import javax.inject.Singleton
+
+@Module
+object PinsModule {
```
to be honest, you can safely combine most part of your dagger modules into a single one. Consider avoiding any extra code

enginire4ek replied:
Fixed

102Rudy commented on data/src/androidTest/java/com/example/pinterbest/data/ExampleInstrumentedTest.kt:
```diff
@@ -0,0 +1,22 @@
+ * See [testing documentation](http://d.android.com/tools/testing).
+ */
+@RunWith(AndroidJUnit4::class)
+class ExampleInstrumentedTest {
```
let's remove that

enginire4ek replied:
Removed

102Rudy commented on data/src/main/java/com/example/pinterbest/data/models/OwnProfile.kt:
```diff
@@ -0,0 +1,28 @@
+
+data class OwnProfile(
+    @SerializedName("ID")
+    val id: Int,
+    val username: String,
+    val email: String,
+    val avatarLink: String,
+    val following: Int,
+    val followers: Int,
+    val boardsCount: Int,
+    val pinsCount: Int
```
remember about pro guard and code obfuscation

enginire4ek replied:
Fixed

102Rudy commented on data/src/main/java/com/example/pinterbest/data/models/OwnProfile.kt:
```diff
@@ -0,0 +1,28 @@
+import com.example.pinterbest.domain.entities.Profile
+import com.google.gson.annotations.SerializedName
+
+data class OwnProfile(
```
By the name of class I should clearly understand what is the layer (data, domain or presentation) the class belongs to?

enginire4ek replied:
Fixed

102Rudy commented on data/src/main/java/com/example/pinterbest/data/repository/AuthRepositoryImpl.kt:
```diff
@@ -0,0 +1,87 @@
+        }.flowOn(Dispatchers.IO)
+
+    override fun saveCookieToPreferences(cookie: String) {
+        SessionRepositoryImpl(preferences).saveSession(cookie)
```
SessionRepositoryImpl must be provided outside of class via constructor. Creating any dependencies inside it's a bad practice

enginire4ek replied:
Fixed

102Rudy commented on data/src/main/java/com/example/pinterbest/data/repository/AuthRepositoryImpl.kt:
```diff
@@ -0,0 +1,87 @@
+        emit(Result.Loading)
+        try {
+            val response = ApiClient().getInstance().getClient().getAuthCheck(
+                SessionRepositoryImpl(preferences).authProvider() ?: ""
```
SessionRepositoryImpl must be provided outside of class via constructor. Creating any dependencies inside it's a bad practice

enginire4ek replied:
Fixed

102Rudy commented on data/src/main/java/com/example/pinterbest/data/repository/PinsRepositoryImpl.kt:
```diff
@@ -0,0 +1,50 @@
+        try {
+            emit(Result.Loading)
+            val pins = ApiClient().getInstance().getClient().getPinFeed(
+                SessionRepositoryImpl(preferences).authProvider() ?: ""
```
SessionRepositoryImpl must be provided outside of class via constructor. Creating any dependencies inside it's a bad practice

enginire4ek replied:
Fixed

102Rudy commented on data/src/main/java/com/example/pinterbest/data/repository/ProfileRepositoryImpl.kt:
```diff
@@ -0,0 +1,37 @@
+        emit(Result.Loading)
+        try {
+            val profile = ApiClient().getInstance().getClient().getProfile(
+                SessionRepositoryImpl(preferences).authProvider() ?: ""
```
SessionRepositoryImpl must be provided outside of class via constructor. Creating any dependencies inside it's a bad practice

enginire4ek replied:
Fixed

102Rudy commented on data/src/test/java/com/example/pinterbest/data/ExampleUnitTest.kt:
```diff
@@ -0,0 +1,16 @@
+ *
+ * See [testing documentation](http://d.android.com/tools/testing).
+ */
+class ExampleUnitTest {
```
may be safely removed

enginire4ek replied:
Removed

102Rudy commented on domain/build.gradle:
```diff
@@ -0,0 +1,19 @@
+}
+
+java {
+    sourceCompatibility = JavaVersion.VERSION_1_7
+    targetCompatibility = JavaVersion.VERSION_1_7
```
why not VERSION_1_8 ?

enginire4ek replied:
Fixed

102Rudy commented on domain/src/main/java/com/example/pinterbest/domain/common/CustomExceptions.kt:
```diff
@@ -0,0 +1,25 @@
+class UserNotFoundException : IllegalStateException()
+
+object ErrorMessage {
+    const val INVALID_DATA = 400
+    const val WRONG_PASSWORD = 401
+    const val ALREADY_AUTHORIZED = 403
+    const val USER_NOT_FOUND = 404
+    const val USER_EXISTS = 409
+
+    val ErrorMap = mapOf(
+        INVALID_DATA to InvalidDataException(),
+        WRONG_PASSWORD to WrongPasswordException(),
+        ALREADY_AUTHORIZED to AlreadyAuthorizedException(),
+        USER_NOT_FOUND to UserNotFoundException(),
+        USER_EXISTS to UserExistsException(),
+    )
```
these codes and mapping should be in data layer

enginire4ek replied:
Fixed

102Rudy commented on domain/src/main/java/com/example/pinterbest/domain/usecases/GetCheckAuthUseCase.kt:
```diff
@@ -0,0 +1,10 @@
+class GetCheckAuthUseCase @Inject constructor(
+    private val authRepository: AuthRepository
+) {
+    suspend operator fun invoke() = authRepository.getCheckAuth()
```
nit: marking the fun as `operator` it's not necessarily

enginire4ek replied:
Convenient use of use-cases and readability of the code. Let's leave it like this for now

102Rudy commented on domain/src/main/java/com/example/pinterbest/domain/usecases/PostLogInUseCase.kt:
```diff
@@ -0,0 +1,11 @@
+import com.example.pinterbest.domain.repositories.AuthRepository
+import javax.inject.Inject
+
+class PostLogInUseCase @Inject constructor(
```
I think you can safely combine some use cases into the single one to avoid boilerplate. It's really unnecessary to create a separate use case for every action

enginire4ek replied:
For more clarity, I decided to make this separation of the use-cases for now. Perhaps with further expansion of the application, the functionality will be combined

102Rudy commented on presentation/src/main/java/com/example/pinterbest/presentation/MainActivity.kt:
```diff
@@ -64,26 +71,29 @@ class MainActivity : AppCompatActivity(), NavigationView.OnNavigationItemSelecte
+                    if (result is Result.Error) {
+                        if (result.exception in ErrorMessage.ErrorMap.values) {
+                            navGraph.startDestination = R.id.loginFragment
+                            navController.graph = navGraph
+                            binding.root.alpha = 1.0F
+                            setContentView(binding.root)
```
can be safely moved out of `if`

enginire4ek replied:
Fixed

enginire4ek committed:
Task 4. Updates
- Several Dagger modules are combined into one
- Deleted ExampleTests
- In domain-layer updated JavaVersion.VERSION_1_8
- Dependency injection fixed in repositories

enginire4ek committed:
Task 4. Updates
- Fixed Exception's files

enginire4ek committed:
Task 4. Updates
- Linter fixes
